### DISCUSSION TITLE
How to check if key exists in map (for a map with many items)

### DISCUSSION THREAD

**Discussion by Roohiya** [cue-lang/cue#1626] [General] [answered]
```
"Items": {
     "item1": {fields...}
     "item2": {fields...}
     ...
}
```

I would like to check if a cue value exists as a key in the above map, and this code works perfectly
```#item: string & or([ for key, _ in services.Entities {key}])```

However since this is imperative, and we have still get all keys and loop through the keys it is not ideal performance wise with 0(n) complexity. It takes very long with 200++ items in the map.

Is there a better way to validate if key exists in map without a for a loop with O(1) time complexity?

**Comment by eonpatapon** [comment, marked as answer]
What may cause a performance issue is that you are generating a big disjunction with `or`, more than the `for` comprehension itself.

A possibility would be to define another field which does some kind of assertion:

```cue
items: {
     "item1": 1
     "item2": 2
}

item: string
itemExists: true & items[item] != _|_

item: "item1"
// item: "item3" // will fail
```

https://cuelang.org/play/?id=SI8GGu0F3WI#cue@export@cue
